samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,295 @@
+
+from samcli.commands._utils.template import get_template_data
+from samcli.commands.build.build_context import BuildContext
+from samcli.commands.deploy.deploy_context import DeployContext
+from samcli.commands.package.package_context import PackageContext
+from samcli.lib.providers.sam_stack_provider import is_local_path
+from samcli.lib.utils.resources import (
+    AWS_SERVERLESS_FUNCTION,
+    AWS_LAMBDA_FUNCTION,
+    AWS_SERVERLESS_LAYERVERSION,
+    AWS_LAMBDA_LAYERVERSION,
+    AWS_SERVERLESS_API,
+    AWS_APIGATEWAY_RESTAPI,
+    AWS_SERVERLESS_HTTPAPI,
+    AWS_APIGATEWAY_V2_API,
+    AWS_SERVERLESS_STATEMACHINE,
+    AWS_STEPFUNCTIONS_STATEMACHINE,
+    SYNCABLE_RESOURCES,
+)
+from samcli.yamlhelper import yaml_parse
+
+LOG = logging.getLogger(__name__)
+
+
+class InfraSyncExecutor:
+    """
+    Executor for infra sync that contains skip logic when template is not changed
+    """
+
+    _build_context: BuildContext
+    _package_context: PackageContext
+    _deploy_context: DeployContext
+
+    def __init__(self, build_context: BuildContext, package_context: PackageContext, deploy_context: DeployContext):
+        """Constructs the sync for infra executor.
+
+        Parameters
+        ----------
+        build_context : BuildContext
+            BuildContext
+        package_context : PackageContext
+            PackageContext
+        deploy_context : DeployContext
+            DeployContext
+        """
+        self._build_context = build_context
+        self._package_context = package_context
+        self._deploy_context = deploy_context
+
+        session = Session(profile_name=self._deploy_context.profile, region_name=self._deploy_context.region)
+        self._cfn_client = session.client("cloudformation")
+        self._s3_client = session.client("s3")
+
+    def _compare_templates(self, local_template_path: str, stack_name: str) -> bool:
+        """
+        Recursively conpares two templates, including the nested templates referenced inside
+
+        Parameters
+        ----------
+        local_template_path : str
+            The local template location
+        stack_name : str
+            The CloudFormation stack name that the template is deployed to
+
+        Returns
+        -------
+        bool
+            Returns True if two templates are identical
+            Returns False if two templates are different
+        """
+        try:
+            # If the customer template uses a nested stack with location/template URL in S3
+            if local_template_path.startswith("https://"):
+                parsed_s3_location = re.search(r"https:\/\/[^/]*\/([^/]*)\/(.*)", local_template_path)
+                s3_bucket = parsed_s3_location.group(1)
+                s3_key = parsed_s3_location.group(2)
+                s3_object = self._s3_client.get_object(Bucket=s3_bucket, Key=s3_key)
+                current_template = yaml_parse(s3_object.get("Body").read().decode("utf-8"))
```

Comment:
Yes, we do. This is needed because the nested template location can be a non-local path. Even though we can ensure that the parent template path is local, we still need this logic for nested templates